pkg/changelog/changelog.go

```diff
@@ -1,224 +1,204 @@
-// the outside world as it is not thread-safe.
+// MaxEntries represents the maximum number of entries that can be stored for a given kind of entry.
+type MaxEntries uint8
 
-type Changelog[T comparable] struct {
-	changes    []item[T]              // list of changes
-	timestamps map[time.Time]struct{} // set of timestamps (used to avoid duplicates)
-	maxSize    int                    // maximum amount of changes to keep track of
-}
+// EntryFlagsMap is a mapping from `MemberKind` to `MaxEntries`.
+type EntryFlagsMap map[MemberKind]MaxEntries
 
-// NewChangelog creates a new changelog.
-func NewChangelog[T comparable](maxSize int) *Changelog[T] {
-	return &Changelog[T]{
-		changes:    []item[T]{},
-		timestamps: map[time.Time]struct{}{},
-		maxSize:    maxSize,
-	}
+// entry is an internal structure representing a single change in the Entries.
+// It includes the kind of the entry, the timestamp of the change, and the value of the change.
+type entry[T comparable] struct {
+	k     MemberKind // Kind of the member, used to categorize the entry.
+	t     time.Time  // Timestamp of when the change occurred.
+	value T          // Value of the change.
 }
 
-// Getters
-
-// GetCurrent: Observation on single element changelog.
-//
-// If there's one element in the changelog, after the loop, left would be set to 1 if the single
-// timestamp is before the targetTime, and 0 if it's equal or after.
-//
-// BEFORE: If the single timestamp is before the targetTime, when we return
-// clv.changes[left-1].value, returns clv.changes[0].value, which is the expected behavior.
-//
-// AFTER: If the single timestamp is equal to, or after the targetTime, the current logic would
-// return a "zero" value because of the condition if left == 0.
-//
-// We need to find the last change that occurred before or exactly at the targetTime. The binary
-// search loop finds the position where a new entry with the targetTime timestamp would be inserted
-// to maintain chronological order:
-//
-// This position is stored in "left".
-//
-// So, to get the last entry that occurred before the targetTime, we need to access the previous
-// position, which is left-1.
+// Entries is the main structure that manages a list of changes (entries).
+// It keeps track of the configured flags for each kind of entry and the list of recorded entries.
 //
-// GetCurrent returns the latest value of the changelog.
-func (clv *Changelog[T]) GetCurrent() T {
-	if len(clv.changes) == 0 {
-		return returnZero[T]()
-	}
-
-	return clv.changes[len(clv.changes)-1].value
+// ATTENTION: You should use Entries within a struct and provide methods to access it,
+// coordinating access through your struct mutexes. DO NOT EXPOSE the Entries object directly to
+// the outside world as it is not thread-safe.
+type Entries[T comparable] struct {
+	entryFlagsMap EntryFlagsMap // Configuration map defining flags for each member kind.
+	entries       []entry[T]    // List of recorded entries.
```

Comment:
I am not sure at what length of entries this would be less optimal than a map. I guess you have made a comparison between this and (intuitively) a map from MemberKind to a time ordered slice? 

Alternatively, instead of a map, since we know at instantiation time the exact count of MemberKind, I think we can create a slice with known access indices per MemberKind (and I think these can just be the enum id). Then each index points to the entry array. This would also mean that we can remove the MemberKind metadata from the `entry[T]` object itself.

Reply:
> I am not sure at what length of entries this would be less optimal than a map. I guess you have made a comparison between this and (intuitively) a map from MemberKind to a time ordered slice?

As this effort focused on reducing the memory footprint, there's a chance that after some size the slice becomes less optimized than a map. But I believe it should be a larger slice, since the hash computation of the map will be more expensive than an iteration which don't trigger a cache miss. The benchmark that has a test case with 15 entries is verifiable anyway, so if we need in the future to use Entries to hold so many changes, we should start by creating incremental size test cases to detect the threshold.

> Alternatively, instead of a map, since we know at instantiation time the exact count of MemberKind, I think we can create a slice with known access indices per MemberKind (and I think these can just be the enum id). Then each index points to the entry array. This would also mean that we can remove the MemberKind metadata from the entry[T] object itself as its now defined through the access index.

This sounds promising, give me some time to check it. 👍🏼

Reply:
We can convert it to a slice, but MemberKind metadata is required to set and get the entry of specific kind. Recall that the `entries` is flat for all member kinds.

Reply:
I'm suggesting a slice of slices. so the subslice is already accessed under the assumption of some MemberKind which is guaranteed by the access index.

Reply:
It works, but doing that we will increase again the metadata in one pointer (8 bytes) + slice head (24 bytes) per kind and certainly we'll have more cache misses.

Reply:
We continued this discussion offline and as @NDStrahilevitz suggested I improved it to use slices instead of maps in the Entries creation what improved the cpu time. That's really good. Tks for the tip, Nadav.

I've benchmarked as well using slice of slices (inner as a member kind one) instead of a single flat. It indeed improved cpu time but did increase the memory footprint as well. Said that, for our specific use case, this current implementation serve us well enough for now.

Reply:
https://github.com/aquasecurity/tracee/pull/4384#issuecomment-2515377613